internal/crypto/cms/signed.go

```diff
@@ -0,0 +1,326 @@
+	var lastErr error
+	for _, signerInfo := range d.SignerInfos {
+		if signerInfo.Version != 1 {
+			// Only IssuerAndSerialNumber is supported currently
+			lastErr = VerificationError{Message: fmt.Sprintf("invalid signer info version: only version 1 is supported; got %d", signerInfo.Version)}
+			continue
+		}
+
+		signingCertificate := d.GetCertificate(signerInfo.SignerIdentifier)
+		if signingCertificate == nil {
+			lastErr = ErrCertificateNotFound
+			continue
+		}
+
+		cert, err := d.verify(&signerInfo, signingCertificate, &opts)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+		thumbprint, err := hashutil.ComputeHash(crypto.SHA256, cert.Raw)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+		verifiedSignerMap[hex.EncodeToString(thumbprint)] = cert
+	}
+	if len(verifiedSignerMap) == 0 {
+		return nil, lastErr
+	}
+
+	verifiedSigningCertificates := make([]*x509.Certificate, 0, len(verifiedSignerMap))
+	for _, cert := range verifiedSignerMap {
+		verifiedSigningCertificates = append(verifiedSigningCertificates, cert)
+	}
+	return verifiedSigningCertificates, nil
+}
+
+// VerifySigner verifies the signerInfo against the signingCertificate.
+//
+// SignedData's certificates field is optional, so this function can be used
+// to verify the signerInfo without the certificates field. Or the user doesn't
+// trust the certificates of SignedData and the signer identifier of
+// the SignerInfo (they are unsigned fields), this function can be used to
+// verify the signerInfo against the user provided signingCertificate.
+//
+// References:
+//   - RFC 5652 5   Signed-data Content Type
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+//
+// WARNING: this function doesn't do any revocation checking.
+func (d *ParsedSignedData) VerifySigner(signerInfo *SignerInfo, signingCertificate *x509.Certificate, opts x509.VerifyOptions) (*x509.Certificate, error) {
+	if signerInfo.Version != 1 {
+		// Only IssuerAndSerialNumber is supported currently
+		return nil, VerificationError{Message: fmt.Sprintf("invalid signer info version: only version 1 is supported; got %d", signerInfo.Version)}
+	}
+
+	intermediates := x509.NewCertPool()
+	for _, cert := range d.Certificates {
+		intermediates.AddCert(cert)
+	}
+	opts.Intermediates = intermediates
+
+	if signingCertificate == nil {
+		// user didn't provide signing certificate, find it from the signed data
+		// certificates
+		signingCertificate = d.GetCertificate(signerInfo.SignerIdentifier)
+		if signingCertificate == nil {
+			return nil, ErrCertificateNotFound
+		}
+	} else {
+		// user provided signing certificate must match the one in signer info
+		if !bytes.Equal(signingCertificate.RawIssuer, signerInfo.SignerIdentifier.Issuer.FullBytes) || signingCertificate.SerialNumber.Cmp(signerInfo.SignerIdentifier.SerialNumber) != 0 {
+			return nil, SyntaxError{Message: "signing certificate does not match signer info"}
+		}
+	}
```

Comment:
For `VerifySigner`, we require caller to provide a cert.
```suggestion
	if signingCertificate == nil {
		return nil, errors.New("verify signer: signingCertificate cannot be nil")
	}
	// user provided signing certificate must match the one in signer info
	if !bytes.Equal(signingCertificate.RawIssuer, signerInfo.SignerIdentifier.Issuer.FullBytes) || signingCertificate.SerialNumber.Cmp(signerInfo.SignerIdentifier.SerialNumber) != 0 {
		return nil, SyntaxError{Message: "signing certificate does not match signer info"}
	}
```

Reply:
Updated.